md2man/roff.go

```diff
@@ -1,252 +1,281 @@
+				r.ListCounters[len(r.ListCounters)-1]++
+			} else if node.ListFlags&blackfriday.ListTypeDefinition != 0 {
+				// state machine for handling terms and following definitions
+				// since blackfriday does not distinguish them properly, nor
+				// does it seperate them into separate lists as it should
+				if !r.defineTerm {
+					out(w, arglistTag)
+					r.defineTerm = true
+				} else {
+					r.defineTerm = false
+				}
+			} else {
+				out(w, ".IP \\(bu 2\n")
+			}
+		} else {
+			out(w, "\n")
+		}
+	case blackfriday.CodeBlock:
+		out(w, codeTag)
+		escapeSpecialChars(w, node.Literal)
+		out(w, codeCloseTag)
+	case blackfriday.Table:
+		if entering {
+			out(w, tableStart)
+			//call walker to count cells (and rows?) so format section can be produced
+			columns := countColumns(node)
+			out(w, strings.Repeat("l ", columns)+"\n")
+			out(w, strings.Repeat("l ", columns)+".\n")
+		} else {
+			out(w, tableEnd)
+		}
+	case blackfriday.TableCell:
+		var (
+			start, end string
+		)
+		if node.IsHeader {
+			start = codespanTag
+			end = codespanCloseTag
+		}
+		if entering {
+			if node.Prev.Type == blackfriday.TableCell {
+				out(w, "\t"+start)
+			}
+		} else {
+			out(w, end)
+		}
+	case blackfriday.TableHead:
+	case blackfriday.TableBody:
+		// no action as cell entries do all the nroff formatting
+		return blackfriday.GoToNext
+	case blackfriday.TableRow:
+		out(w, "\n")
+	default:
+		fmt.Fprintf(os.Stderr, "WARNING: go-md2man does not handle node type "+node.Type.String())
 	}
 
-	out.Write(body)
-	out.WriteString("\n.TE\n")
+	return blackfriday.GoToNext
 }
 
-func (r *roffRenderer) TableRow(out *bytes.Buffer, text []byte) {
-	if out.Len() > 0 {
-		out.WriteString("\n")
-	}
-	out.Write(text)
-}
+// because roff format requires knowing the column count before outputting any table
+// data we need to walk a table tree and count the columns
+func countColumns(node *blackfriday.Node) int {
+	var columns int
 
-func (r *roffRenderer) TableHeaderCell(out *bytes.Buffer, text []byte, align int) {
-	if out.Len() > 0 {
-		out.WriteString("\t")
-	}
-	if len(text) == 0 {
-		text = []byte{' '}
-	}
-	out.Write([]byte("\\fB\\fC" + string(text) + "\\fR"))
+	node.Walk(func(node *blackfriday.Node, entering bool) blackfriday.WalkStatus {
+		switch node.Type {
+		case blackfriday.TableRow:
+			if !entering {
+				return blackfriday.Terminate
+			}
+		case blackfriday.TableCell:
+			columns++
+		default:
+			return blackfriday.GoToNext
+		}
+		return blackfriday.Terminate
+	})
+	return columns
 }
 
-func (r *roffRenderer) TableCell(out *bytes.Buffer, text []byte, align int) {
-	if out.Len() > 0 {
-		out.WriteString("\t")
-	}
-	if len(text) > 30 {
-		text = append([]byte("T{\n"), text...)
-		text = append(text, []byte("\nT}")...)
-	}
-	if len(text) == 0 {
-		text = []byte{' '}
+func out(w io.Writer, output string) {
+	if _, err := io.WriteString(w, output); err != nil {
+		fmt.Fprintf(os.Stderr, "ERROR: Couldn't output string to writer: %v\n", err)
```

Comment:
Maybe instead of printing to stderr we can take a logger object?

Reply:
Yeah.. wasn't sure what was best here. More broadly, the `blackfriday.Renderer` interface definition gives us no way to pass back errors, so other than output or logging in some way, not sure how to error out. I believe you can tell the parser to stop traversing markdown at the current node (effectively not continue and exit), but it isn't seen as an error situation.

Reply:
Yeah, this is unfortunate. I'm suprised they didn't decide to allow renderers to return an error, but admittedly I didn't get any time to look into the interface when it was being built. 😞